kernel/src/engine/arrow_expression/in_list.rs

```diff
@@ -0,0 +1,566 @@
+use std::cmp::Ordering;
+use std::sync::Arc;
+
+use arrow_array::cast::AsArray;
+use arrow_array::{types::*, ArrowNumericType, PrimitiveArray};
+use arrow_array::{Array, ArrayRef, BooleanArray, RecordBatch};
+use arrow_cast::cast;
+use arrow_ord::comparison::{in_list, in_list_utf8};
+use arrow_schema::{DataType as ArrowDataType, TimeUnit};
+use paste::paste;
+
+use super::{evaluate_expression, extract_column};
+use crate::error::{DeltaResult, Error};
+use crate::expressions::{ArrayData, Expression, Scalar, VariadicOperator};
+use crate::predicates::PredicateEvaluatorDefaults;
+use crate::schema::PrimitiveType;
+
+pub(super) fn eval_in_list(
+    batch: &RecordBatch,
+    left: &Expression,
+    right: &Expression,
+) -> Result<ArrayRef, Error> {
+    use Expression::*;
+
+    let result = match (left, right) {
+        (Literal(Scalar::Null(_)), Column(_) | Literal(Scalar::Array(_))) => {
+            // Searching any in-list for NULL always returns NULL -- no need to actually search
+            BooleanArray::new_null(batch.num_rows())
+        }
+        (Literal(lit), Literal(Scalar::Array(ad))) => {
+            // Search the literal in-list once and then replicate the returned single-row result
+            let exists = is_in_list(ad, Some(Some(lit.clone()))).iter().next();
+            BooleanArray::from(vec![exists.flatten(); batch.num_rows()])
+        }
+        (Literal(_) | Column(_), Column(_)) => {
+            let right_arr = evaluate_expression(right, batch, None)?;
+            let list_field = match right_arr.data_type() {
+                ArrowDataType::List(list_field) => list_field,
+                ArrowDataType::LargeList(list_field) => list_field,
+                // TODO: LargeListView - not fully supported by arrow yet
+                _ => {
+                    return Err(Error::invalid_expression(format!(
+                        "Expected right hand side to be list column, got: {:?}",
+                        right_arr.data_type()
+                    )));
+                }
+            };
+
+            let left_arr =
+                evaluate_expression(left, batch, Some(&list_field.data_type().try_into()?))?;
+            // we should at least cast string / large string arrays to the same type, and may as well see if
+            // we can cast other types.
+            let left_arr = if left_arr.data_type() == list_field.data_type() {
+                left_arr
+            } else {
+                cast(left_arr.as_ref(), list_field.data_type()).map_err(Error::generic_err)?
+            };
```

Comment:
while technically not really the right place to cast values, it may actually help us when we start to support type widening. Then again, there values should really be cast when reading the data.

Nonetheless it makes the code simpler if we always try to cast, and I thought its Ok if we do this here?

Reply:
looking into the [docs](https://docs.rs/arrow-cast/53.4.1/arrow_cast/cast/fn.cast_with_options.html#behavior) it seems we should not be using this function, as it tries to also parse strings into literals.

So our need here is casting the various physical representations for the same type.

We could do a pre-factor PR that just writes a simple function to do allowed casts as we will need that for type widening anyway? IIRC we need some casting to pass tests ...